presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java

```diff
@@ -0,0 +1,98 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.sketch.kll;
+
+import com.facebook.presto.common.array.ObjectBigArray;
+import com.facebook.presto.operator.aggregation.sketch.theta.ThetaSketchStateFactory;
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorStateMetadata;
+import org.apache.datasketches.kll.KllItemsSketch;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+@AccumulatorStateMetadata(stateFactoryClass = KllSketchStateFactory.class, stateSerializerClass = KllSketchStateSerializer.class)
+public interface KllSketchAggregationState
+        extends AccumulatorState
+{
+    @Nullable
+    <T> KllItemsSketch<T> getSketch();
+
+    <T> void setSketch(KllItemsSketch<T> sketch);
+
+    class Single
+            implements KllSketchAggregationState
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(Single.class).instanceSize();
+        @Nullable
+        private KllItemsSketch sketch;
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public <T> KllItemsSketch<T> getSketch()
+        {
+            return sketch;
+        }
+
+        @Override
+        public <T> void setSketch(KllItemsSketch<T> sketch)
+        {
+            this.sketch = sketch;
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            if (sketch != null) {
+                return sketch.getSerializedSizeBytes() + INSTANCE_SIZE;
+            }
+            return INSTANCE_SIZE;
+        }
+    }
+
+    class Grouped
+            extends AbstractGroupedAccumulatorState
+            implements KllSketchAggregationState
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(ThetaSketchStateFactory.GroupedThetaSketchState.class).instanceSize();
+        private final ObjectBigArray<KllItemsSketch> sketches = new ObjectBigArray<>();
+
+        @Override
+        public <T> KllItemsSketch<T> getSketch()
+        {
+            return sketches.get(getGroupId());
+        }
+
+        @Override
+        public <T> void setSketch(KllItemsSketch<T> sketch)
+        {
+            sketches.set(getGroupId(), requireNonNull(sketch, "kll sketch is null"));
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return sketches.sizeOf() + INSTANCE_SIZE;
```

Comment:
The problem with this is it only estimates the size of the list, not the size of the memory used in aggregate.  It needs to be `SIZE_OF(KLLSketch) * sketches.sizeOf() + INSTANCE_SIZE`.

One way of doing this is to keep a running total in the grouped aggregation state, and increment/decrement it whenever sketches are added.

Reply:
Fixed